File: server/decoder/coords.go
package decoder

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"strings"
)

type CoordResolver interface {
	Convert(rawLat, rawLng uint32) (float64, float64)
	Mode() CoordMode
}

// RelativeResolver normalizes to 0-1 range
type RelativeResolver struct {
	minLat, maxLat, minLng, maxLng float64
}

func NewRelativeResolver(minLat, maxLat, minLng, maxLng uint32) *RelativeResolver {
	return &RelativeResolver{
		minLat: float64(minLat), maxLat: float64(maxLat),
		minLng: float64(minLng), maxLng: float64(maxLng),
	}
}

func (r *RelativeResolver) Convert(rawLat, rawLng uint32) (float64, float64) {
	lat := (float64(rawLat) - r.minLat) / (r.maxLat - r.minLat)
	lng := (float64(rawLng) - r.minLng) / (r.maxLng - r.minLng)
	return math.Max(0, math.Min(1, lat)), math.Max(0, math.Min(1, lng))
}

func (r *RelativeResolver) Mode() CoordMode { return CoordRelative }

// WGS84Resolver applies a linear transform to WGS84
type WGS84Resolver struct {
	LatScale  float64
	LatOffset float64
	LngScale  float64
	LngOffset float64
}

func (r *WGS84Resolver) Convert(rawLat, rawLng uint32) (float64, float64) {
	lat := float64(rawLat)*r.LatScale + r.LatOffset
	lng := float64(rawLng)*r.LngScale + r.LngOffset
	return lat, lng
}

func (r *WGS84Resolver) Mode() CoordMode { return CoordWGS84 }

// MapMeta is the metadata from the .txt sidecar file next to the .db
type MapMeta struct {
	OK            bool    `json:"OK"`
	CenterOK      bool    `json:"CenterOK"`
	CenterLat     float64 `json:"CenterLat"`
	CenterLng     float64 `json:"CenterLng"`
	MaxNativeZoom int     `json:"MaxNativeZoom"`
	GratCR        int     `json:"GratCR"`
	GratLatBegin  float64 `json:"GratLatBegin"`
	GratLatSpace  float64 `json:"GratLatSpace"`
	GratLngBegin  float64 `json:"GratLngBegin"`
	GratLngSpace  float64 `json:"GratLngSpace"`
}

// LoadMapMeta reads the .txt sidecar file for a .db file.
func LoadMapMeta(dbPath string) (*MapMeta, error) {
	txtPath := strings.TrimSuffix(dbPath, ".db") + ".txt"
	data, err := os.ReadFile(txtPath)
	if err != nil {
		return nil, err
	}
	var meta MapMeta
	if err := json.Unmarshal(data, &meta); err != nil {
		return nil, err
	}
	if !meta.OK || !meta.CenterOK {
		return nil, nil
	}
	return &meta, nil
}

// AutoDetectCoords determines the coordinate resolver.
// If a .txt sidecar file exists, it uses the center coordinates to compute
// the offset between raw database values and WGS84.
func AutoDetectCoords(db *sql.DB, dbPath string) (CoordResolver, CoordMode, error) {
	var minLat, maxLat, minLng, maxLng uint32
	err := scanCoordBounds(db, &minLat, &maxLat, &minLng, &maxLng)
	if err != nil {
		return nil, CoordRelative, err
	}

	// If a .txt metadata file exists and has valid data, use the known encoding:
	// raw = (WGS84 + 180) × 1e6  →  WGS84 = raw × 1e-6 - 180
	if meta, err := LoadMapMeta(dbPath); err == nil && meta != nil {
		return &WGS84Resolver{
			LatScale: 1e-6, LatOffset: -180,
			LngScale: 1e-6, LngOffset: -180,
		}, CoordWGS84, nil
	}

	// Fallback: heuristic auto-detection

	// Try: raw * 1e-6 - 180 (common encoding: (WGS84 + 180) × 1e6)
	testLat6 := float64(minLat)*1e-6 - 180
	testLng6 := float64(minLng)*1e-6 - 180
	if testLat6 > -90 && testLat6 < 90 && testLng6 > -180 && testLng6 < 180 {
		return &WGS84Resolver{LatScale: 1e-6, LatOffset: -180, LngScale: 1e-6, LngOffset: -180}, CoordWGS84, nil
	}

	// Try: raw / 1e7 directly
	testLat := float64(minLat) / 1e7
	testLng := float64(minLng) / 1e7
	if testLat > -90 && testLat < 90 && testLng > -180 && testLng < 180 {
		return &WGS84Resolver{LatScale: 1e-7, LatOffset: 0, LngScale: 1e-7, LngOffset: 0}, CoordWGS84, nil
	}

	// Try: lat/1e7, lng/1e7 + 80
	testLng80 := float64(minLng)/1e7 + 80
	if testLat > 18 && testLat < 55 && testLng80 > 73 && testLng80 < 136 {
		return &WGS84Resolver{LatScale: 1e-7, LatOffset: 0, LngScale: 1e-7, LngOffset: 80}, CoordWGS84, nil
	}

	// Try: lat/1e7, lng/1e7 + 90
	testLng90 := float64(minLng)/1e7 + 90
	if testLat > 18 && testLat < 55 && testLng90 > 73 && testLng90 < 136 {
		return &WGS84Resolver{LatScale: 1e-7, LatOffset: 0, LngScale: 1e-7, LngOffset: 90}, CoordWGS84, nil
	}

	// Fallback to relative
	return NewRelativeResolver(minLat, maxLat, minLng, maxLng), CoordRelative, nil
}

func scanCoordBounds(db *sql.DB, minLat, maxLat, minLng, maxLng *uint32) error {
	rows, err := db.Query(`
		SELECT LogData FROM record
		WHERE SrcType=1 AND DataType=1 AND LogData IS NOT NULL
	`)
	if err != nil {
		return err
	}
	defer rows.Close()

	*minLat = math.MaxUint32
	*minLng = math.MaxUint32
	*maxLat = 0
	*maxLng = 0

	found := false
	for rows.Next() {
		var blob []byte
		if err := rows.Scan(&blob); err != nil {
			continue
		}
		units := DecodePositionFrame(blob)
		for _, u := range units {
			if u.RawLat == 0 && u.RawLng == 0 {
				continue
			}
			found = true
			if u.RawLat < *minLat {
				*minLat = u.RawLat
			}
			if u.RawLat > *maxLat {
				*maxLat = u.RawLat
			}
			if u.RawLng < *minLng {
				*minLng = u.RawLng
			}
			if u.RawLng > *maxLng {
				*maxLng = u.RawLng
			}
		}
	}
	if !found {
		return fmt.Errorf("no position data found in database")
	}
	return nil
}

File: src/engine/coords.ts
// ── Coordinate detection ported from Go decoder/coords.go ──

import type { CoordMode, MapMeta } from './types';
import type { Database } from 'sql.js';
import { decodePositionFrame } from './decoder';

export interface CoordResolver {
  convert(rawLat: number, rawLng: number): [number, number];
  mode(): CoordMode;
}

class RelativeResolver implements CoordResolver {
  private minLat: number;
  private maxLat: number;
  private minLng: number;
  private maxLng: number;

  constructor(minLat: number, maxLat: number, minLng: number, maxLng: number) {
    this.minLat = minLat;
    this.maxLat = maxLat;
    this.minLng = minLng;
    this.maxLng = maxLng;
  }

  convert(rawLat: number, rawLng: number): [number, number] {
    const rangeLat = this.maxLat - this.minLat;
    const rangeLng = this.maxLng - this.minLng;
    const lat = rangeLat > 0 ? (rawLat - this.minLat) / rangeLat : 0;
    const lng = rangeLng > 0 ? (rawLng - this.minLng) / rangeLng : 0;
    return [Math.max(0, Math.min(1, lat)), Math.max(0, Math.min(1, lng))];
  }

  mode(): CoordMode { return 'relative'; }
}

class WGS84Resolver implements CoordResolver {
  private latScale: number;
  private latOffset: number;
  private lngScale: number;
  private lngOffset: number;

  constructor(latScale: number, latOffset: number, lngScale: number, lngOffset: number) {
    this.latScale = latScale;
    this.latOffset = latOffset;
    this.lngScale = lngScale;
    this.lngOffset = lngOffset;
  }

  convert(rawLat: number, rawLng: number): [number, number] {
    return [rawLat * this.latScale + this.latOffset, rawLng * this.lngScale + this.lngOffset];
  }

  mode(): CoordMode { return 'wgs84'; }
}

/** Scan all DT=1 records to find raw coordinate bounds. */
function scanCoordBounds(db: Database): { minLat: number; maxLat: number; minLng: number; maxLng: number } | null {
  const stmt = db.prepare(`SELECT LogData FROM record WHERE SrcType=1 AND DataType=1 AND LogData IS NOT NULL`);

  let minLat = 0xFFFFFFFF, maxLat = 0, minLng = 0xFFFFFFFF, maxLng = 0;
  let found = false;

  while (stmt.step()) {
    const blob = stmt.get()[0] as Uint8Array;
    if (!blob) continue;
    for (const u of decodePositionFrame(blob)) {
      if (u.rawLat === 0 && u.rawLng === 0) continue;
      found = true;
      if (u.rawLat < minLat) minLat = u.rawLat;
      if (u.rawLat > maxLat) maxLat = u.rawLat;
      if (u.rawLng < minLng) minLng = u.rawLng;
      if (u.rawLng > maxLng) maxLng = u.rawLng;
    }
  }
  stmt.free();

  return found ? { minLat, maxLat, minLng, maxLng } : null;
}

/**
 * Auto-detect coordinate system using 5 heuristics.
 * Returns [resolver, coordMode].
 */
export function autoDetectCoords(
  db: Database,
  txtContent?: string,
): [CoordResolver, CoordMode] {
  const bounds = scanCoordBounds(db);
  if (!bounds) {
    return [new RelativeResolver(0, 1, 0, 1), 'relative'];
  }

  // Heuristic 1: .txt sidecar metadata — raw = (WGS84 + 180) × 1e6
  if (txtContent) {
    try {
      const meta: MapMeta = JSON.parse(txtContent);
      if (meta.OK && meta.CenterOK) {
        return [new WGS84Resolver(1e-6, -180, 1e-6, -180), 'wgs84'];
      }
    } catch { /* invalid JSON, fall through */ }
  }

  // Heuristic 2: raw × 1e-6 - 180
  const testLat6 = bounds.minLat * 1e-6 - 180;
  const testLng6 = bounds.minLng * 1e-6 - 180;
  if (testLat6 > -90 && testLat6 < 90 && testLng6 > -180 && testLng6 < 180) {
    return [new WGS84Resolver(1e-6, -180, 1e-6, -180), 'wgs84'];
  }

  // Heuristic 3: raw / 1e7
  const testLat = bounds.minLat / 1e7;
  const testLng = bounds.minLng / 1e7;
  if (testLat > -90 && testLat < 90 && testLng > -180 && testLng < 180) {
    return [new WGS84Resolver(1e-7, 0, 1e-7, 0), 'wgs84'];
  }

  // Heuristic 4: lat/1e7, lng/1e7 + 80 (China)
  const testLng80 = bounds.minLng / 1e7 + 80;
  if (testLat > 18 && testLat < 55 && testLng80 > 73 && testLng80 < 136) {
    return [new WGS84Resolver(1e-7, 0, 1e-7, 80), 'wgs84'];
  }

  // Heuristic 5: lat/1e7, lng/1e7 + 90
  const testLng90 = bounds.minLng / 1e7 + 90;
  if (testLat > 18 && testLat < 55 && testLng90 > 73 && testLng90 < 136) {
    return [new WGS84Resolver(1e-7, 0, 1e-7, 90), 'wgs84'];
  }

  // Fallback: relative coordinates
  return [new RelativeResolver(bounds.minLat, bounds.maxLat, bounds.minLng, bounds.maxLng), 'relative'];
}

/** Parse .txt sidecar content into MapMeta. */
export function parseMapMeta(txtContent: string): MapMeta | null {
  try {
    const meta: MapMeta = JSON.parse(txtContent);
    if (meta.OK && meta.CenterOK) return meta;
    return null;
  } catch {
    return null;
  }
}
